### DISCUSSION TITLE
Delayed fetching of eclipse-feature dependencies artifacts?

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#2031] [Q&A] [answered]
In a custom plugin I'm writing, I'm accessing projects' dependencies through `MavenProject#getDependencyArtifacts`.

In case of `eclipse-feature` projects, I can see certain artifacts have dummy files wtih the `file not yet available` name.  
I've noticed that in `MavenDependencyInjector`, the `fetch` parameter is set to `true` only for `eclipse-plugin`, why so?

Ultimately my question is, how do I get the correct dependency `Artifact#getFile` in my own plugin?

**Comment by laeubi** [comment, marked as answer]
Features don't have a compile phase so no real dependencies, the question is then what exactly you like to access?
If you want the P2 data you can use `org.eclipse.tycho.core.utils.TychoProjectUtils.getDependencyArtifacts(ReactorProject)`

**Reply by lppedd** [reply]
> the question is then what exactly you like to access

@laeubi good question. I'm scanning `feature.xml` files for required `plugin`s, and I need to copy the actual Artifact JAR files to a temporary directory to let another tool scan them.

Example `feature.xml`:

```
<feature>
   ...
   <plugin id="org.mozilla.javascript" version="..." unpack="false"/>
   <plugin id="org.eclipse.birt.core" version="..." unpack="false"/>
   ...
</feature>
```

In this case I need to access the two artifact's JAR files.  
Best thing would be to avoid any hard reference to Tycho.

**Reply by laeubi** [reply]
You might want to take a look at `org.eclipse.tycho.extras.pde.ListDependenciesMojo`, if you like you might want add a copy-goal if text file do not suffice.

But be aware that the "dependencies" of a feature are not necessarily what you see in the `feature.xml`.

**Reply by lppedd** [reply]
@laeubi that's interesting. I see that internally it uses the `TychoProjectUtils` you mentioned.  
I'll write a `copy` goal as a last resort.

Is there a way to access the artifacts through the standard Maven repository system?
Or is all managed by Tycho classes and I necessarily have to depend on Tycho?

**Reply by laeubi** [reply]
You can try to use the `WorkspaceReader` from the `RepositorySession` to fetch the file

**Reply by lppedd** [reply]
@laeubi this might actually work! I see that at some point `TychoWorkspaceReader` calls `descriptor.getLocation(true /* fetch */)`.  
Now I need to find a utility method to convert `org.apache.maven.artifact.Artifact` to `org.eclipse.aether.artifact.Artifact`.

Edit: `RepositoryUtils.toArtifacts` should do it.

**Reply by lppedd** [reply]
@laeubi Works!  
One can also pass by `RepositorySystem#resolveArtifact`. It's a bit more verbose but it seems a better way.
